src/Model/LazyBSONDocument.php

```diff
@@ -0,0 +1,235 @@
+     *        When given a BSON document, this is treated as input. For arrays
+     *        and objects this constructs a new BSON document using fromPHP.
+     */
+    public function __construct($input = null, ?CodecLibrary $codecLibrary = null)
+    {
+        if ($input === null) {
+            $this->bson = Document::fromPHP([]);
+        } elseif ($input instanceof Document) {
+            $this->bson = $input;
+        } elseif (is_array($input) || is_object($input)) {
+            $this->bson = Document::fromPHP([]);
+
+            foreach ($input as $key => $value) {
+                assert(is_string($key));
+                $this->set[$key] = $value;
+                $this->exists[$key] = true;
+            }
+        } else {
+            throw InvalidArgumentException::invalidType('input', $input, [Document::class, 'array', 'null']);
+        }
+
+        $this->codecLibrary = $codecLibrary ?? new LazyBSONCodecLibrary();
+    }
+
+    /** @return TValue */
+    public function __get(string $property)
+    {
+        $this->readFromBson($property);
+
+        if (isset($this->unset[$property]) || ! $this->exists[$property]) {
+            trigger_error(sprintf('Undefined property: %s', $property), E_USER_WARNING);
+
+            return null;
+        }
+
+        return array_key_exists($property, $this->set) ? $this->set[$property] : $this->read[$property];
+    }
+
+    public function __isset(string $name): bool
+    {
+        // If we've looked for the value, return the cached result
+        if (isset($this->exists[$name])) {
+            return $this->exists[$name];
+        }
+
+        return $this->exists[$name] = $this->bson->has($name);
+    }
+
+    /** @param TValue $value */
+    public function __set(string $property, $value): void
+    {
+        $this->set[$property] = $value;
+        unset($this->unset[$property]);
+        $this->exists[$property] = true;
+    }
+
+    public function __unset(string $name): void
+    {
+        $this->unset[$name] = true;
+        $this->exists[$name] = false;
+        unset($this->set[$name]);
+    }
+
+    /** @return Iterator<string, TValue> */
+    public function getIterator(): CallbackIterator
+    {
+        $itemIterator = new AppendIterator();
+        // Iterate through all fields in the BSON document
+        $itemIterator->append($this->bson->getIterator());
+        // Then iterate over all fields that were set
+        $itemIterator->append(new ArrayIterator($this->set));
+
+        /** @var array<string, bool> $seen */
+        $seen = [];
+
+        return new CallbackIterator(
+            // Skip keys that were unset or handled in a previous iterator
+            new CallbackFilterIterator(
+                $itemIterator,
+                /** @param TValue $current */
+                function ($current, string $key) use (&$seen): bool {
+                    return ! isset($this->unset[$key]) && ! isset($seen[$key]);
+                },
```

Comment:
Rector should have detected this.
```suggestion
                fn ($current, string $key): bool => ! isset($this->unset[$key]) && ! isset($seen[$key]),
```

Reply:
It did - it just isn't part of my workflow yet. This is fixed now.